crates/oxc_minifier/src/compressor.rs

```diff
@@ -1,62 +1,80 @@
 use oxc_allocator::Allocator;
 #[allow(clippy::wildcard_imports)]
 use oxc_ast::{ast::*, AstBuilder};
+use oxc_semantic::{ScopeTree, SemanticBuilder, SymbolTable};
+use oxc_traverse::TraverseCtx;
 
 use crate::{
     ast_passes::{
         Collapse, FoldConstants, RemoveDeadCode, RemoveSyntax, SubstituteAlternateSyntax,
     },
-    CompressOptions,
+    CompressOptions, CompressorPass,
 };
 
 pub struct Compressor<'a> {
+    allocator: &'a Allocator,
     ast: AstBuilder<'a>,
     options: CompressOptions,
 }
 
 impl<'a> Compressor<'a> {
     pub fn new(allocator: &'a Allocator, options: CompressOptions) -> Self {
         let ast = AstBuilder::new(allocator);
-        Self { ast, options }
+        Self { allocator, ast, options }
     }
 
     pub fn build(self, program: &mut Program<'a>) {
+        let (symbols, scopes) = SemanticBuilder::new("", program.source_type)
+            .build(program)
+            .semantic
+            .into_symbol_table_and_scope_tree();
+        self.build_with_symbols_and_scopes(symbols, scopes, program);
+    }
+
+    pub fn build_with_symbols_and_scopes(
+        self,
+        symbols: SymbolTable,
+        scopes: ScopeTree,
+        program: &mut Program<'a>,
+    ) {
+        let mut ctx = TraverseCtx::new(scopes, symbols, self.allocator);
```

Comment:
The goal is to reuse `ctx` between AST passes, implementation needs to keep symbols and scopes in sync.

Reply:
I'll look into how we can do this.